### PR TITLE
feat openid login with google

### PR DESCRIPTION
**Pull request by fredguth**
Today the login works only wit HuggingFace as a provider, this PR opens the possibility of adding other OPENID providers.
- changes `hfUserId`to `providerUserId`
- adds optional email to user type
- changes scope to add `email` as `preferred_username` is  not available obligatory and Google does not provide it
- fix `sha256.ts` import

### CODE REVIEW

**Review by coyotte508 — COMMENTED** [review]
Thanks for the PR!

File: src/lib/server/auth.ts [hunk 1]
```diff
@@ -33,7 +33,7 @@ export function refreshSessionCookie(cookies: Cookies, sessionId: string) {
 	});
 }
 
-export const OIDC_SCOPES = "openid profile";
+export const OIDC_SCOPES = "openid profile email";
```
[the remark this answers is not in the document]
Comment (fredguth) [reply src/lib/server/auth.ts:36]: preferred_username is not required by openid standard.  Google does not implement it (and I imagine other providers also don't).  From the email, I take a username.  That is the only reason I added email to the scope.

File: src/routes/login/callback/updateUser.ts [hunk 2]
```diff
@@ -11,13 +11,15 @@ export async function updateUser(params: {
 	locals: App.Locals;
 	cookies: Cookies;
 }) {
+	console.log("updateUser");
```
[the remark this answers is not in the document]
Comment (fredguth) [reply src/routes/login/callback/updateUser.ts:14]: oops

File: src/routes/login/callback/updateUser.ts [hunk 3]
```diff
@@ -11,13 +11,15 @@ export async function updateUser(params: {
...
-
+	userData["preferred_username"] =
+		userData["preferred_username"] || userData["email"]?.split("@")[0];
 	const {
 		preferred_username: username,
 		name,
 		picture: avatarUrl,
-		sub: hfUserId,
+		sub: providerUserId,
 	} = z
```
[the remark this answers is not in the document]
Comment (fredguth) [reply src/routes/login/callback/updateUser.ts:23]: I did that later on. I can add that to the DB.  So, what will it be?  Should we require email and save it to the db?

File: src/lib/types/User.ts [hunk 4]
```diff
@@ -6,8 +6,9 @@ export interface User extends Timestamps {
 
 	username: string;
```
[the remark this answers is not in the document]
Comment (fredguth) [reply src/lib/types/User.ts:7]: I used the first part of the email (before @) as username for Google.  But I think that is not a good solution.  I agree with all the points of attention of this review and they are all easy to change... if we have a decision on how we want it to be. 

I am not the right person to decide it, though.  I can implement either way.

File: src/lib/server/auth.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (coyotte508) [reply src/lib/server/auth.ts]: This is what I meant: https://github.com/huggingface/chat-ui/pull/250/commits/71cfd9108325a2ed815df122e217c2dac070dca8

File: src/routes/login/callback/updateUser.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (coyotte508) [reply src/routes/login/callback/updateUser.ts]: That's what I meant: https://github.com/huggingface/chat-ui/pull/250/commits/a97f1bdd55c4796cc39262ef597a86ea5dd70e07

Not required. But if present, we save it to DB.

File: src/lib/types/User.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (coyotte508) [reply src/lib/types/User.ts]: Handled in https://github.com/huggingface/chat-ui/pull/250/commits/a97f1bdd55c4796cc39262ef597a86ea5dd70e07

**Comment by fredguth** [comment]
I believe I made the changes from the review.

File: src/routes/login/callback/updateUser.ts [hunk 5]
```diff
@@ -12,12 +12,14 @@ export async function updateUser(params: {
 	cookies: Cookies;
 }) {
 	const { userData, locals, cookies } = params;
+	// when using Google as OpenID provider, remember that it does not provide preferred_username
+	// example fix: userData.preferred_username = userData.email.split("@")[0];
```
Comment (coyotte508) [inline src/routes/login/callback/updateUser.ts:16]: It's fine, the email will be displayed in the frontend if the username doesn't exist :)

**Review by coyotte508 — APPROVED** [review]

**Comment by coyotte508** [comment]
Nice @fredguth ! Thanks for adding this

File: src/routes/login/callback/updateUser.ts [hunk 6]
```diff
@@ -11,13 +11,11 @@ export async function updateUser(params: {
 	locals: App.Locals;
 	cookies: Cookies;
 }) {
-	console.log("updateUser");
 	const { userData, locals, cookies } = params;
-	userData["preferred_username"] =
```
Comment (nilesh-patil) [inline src/routes/login/callback/updateUser.ts]: I believe this is required, the login process keep breaking without adding this for me.

Comment (coyotte508) [reply src/routes/login/callback/updateUser.ts]: Should be fixed by https://github.com/huggingface/chat-ui/commit/002a2a0f588a701e0c0f96e2d6c7aef96a8cc512 , can you test?

**Pull request merged**